File: tools/queue_sim.cpp
// queue_sim — replay a feed with one or more of our orders resting in it, and
// report what each fill model says we would have traded.
//
// This is the phase's differential-test driver: python/reference/queue_sim.py
// does the same thing slowly and obviously, and the two must produce byte-
// identical fill logs. That discipline caught real bugs in phases 2 and 3.
//
// The order of operations per message is load-bearing and is mirrored exactly
// in the Python oracle:
//
//   1. session/halt state is updated first, so all lanes see identical eligibility
//   2. the pre-mutation snapshot is taken (E/C/X/D/U carry only a reference)
//   3. the book is mutated
//   4. the message is resolved and committed to every lane
//   5. price priority is evaluated against the post-mutation book
//
// Usage:
//   queue_sim <feed.gz> --place B:1000000:100:500[:display[:cancel_after]]
//             [--place ...] [--fills out.csv] [--json out.json]
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>

#include "itchbook/book/book.hpp"
#include "itchbook/book/dispatch.hpp"
#include "itchbook/itch/parser.hpp"
#include "itchbook/itch/reader.hpp"
#include "itchbook/sim/queue_model.hpp"
#include "itchbook/sim/resolve.hpp"

using namespace itchbook;
using namespace itchbook::sim;

namespace {

struct Placement {
    Side side = Side::Buy;
    Price4 price = 0;
    uint32_t quantity = 0;
    uint64_t after = 0;      // place once this many messages have been applied
    uint32_t display = 0;
    // Pull the order once this many messages have been applied. Zero means
    // never, which is the default and the right one for a sweep; the
    // leave-one-out oracle needs it, because a replacement order that outlives
    // the order it replaced keeps filling after that order was gone and every
    // model reads as over-filling for a reason that is not the model.
    uint64_t cancel_after = 0;
    uint64_t id = 0;
    bool placed = false;
    bool cancelled = false;
};

const char* trigger_name(Trigger t) {
    switch (t) {
        case Trigger::Execution: return "execution";
        case Trigger::Lock:      return "lock";
        case Trigger::Through:   return "through";
        case Trigger::Naive:     return "naive";
        case Trigger::Taking:    return "taking";
    }
    return "?";
}

struct Lane {
    Model model;
    QueueModel q;
    std::vector<SimFill> fills;
    explicit Lane(Model m, QueueConfig c) : model(m), q(c) {}
};

struct Handler {
    book::Book book{100, 20, 1u << 20};
    std::vector<Lane>* lanes = nullptr;
    std::vector<Placement>* places = nullptr;
    uint64_t index = 0;

    void on_message(char type, const uint8_t* p, uint16_t) {
        // 1. session and halt state, before anything else, so every lane sees
        //    byte-identical eligibility.
        if (type == 'H') {
            const char state = itch::trading_action::state(p);
            for (Lane& l : *lanes) l.q.set_trading_state(state);
        } else if (type == 'S') {
            const char code = itch::system_event::code(p);
            // 'Q' opens market hours, 'M' closes them. Everything else leaves
            // the state as it was.
            if (code == 'Q') {
                for (Lane& l : *lanes) l.q.set_trading_state('T');
            } else if (code == 'M' || code == 'C') {
                for (Lane& l : *lanes) l.q.set_trading_state('\0');
            }
        }

        // 2 + 3. snapshot, then mutate.
        book::PreState pre;
        book::apply_ex(book, type, p, &pre);

        // 4. resolve and commit.
        const Resolved r = resolve(type, p, pre);
        for (Lane& l : *lanes) l.q.commit(r, book, &l.fills);

        // 5. price priority against the post-mutation book.
        const uint64_t ts = itch::timestamp(p);
        for (Lane& l : *lanes) l.q.apply_price_priority(book, ts, &l.fills);

        ++index;

        // Placements happen after the message that reaches their index, using
        // the book as it stands at that moment — arrival-time queue position,
        // never decision-time.
        for (Placement& pl : *places) {
            if (pl.placed || index < pl.after) continue;
            for (Lane& l : *lanes) {
                l.q.place(book, pl.id, pl.side, pl.price, pl.quantity, pl.display, ts);
            }
            pl.placed = true;
        }
        for (Placement& pl : *places) {
            if (!pl.placed || pl.cancelled || pl.cancel_after == 0) continue;
            if (index < pl.cancel_after) continue;
            for (Lane& l : *lanes) l.q.cancel(pl.id);
            pl.cancelled = true;
        }
    }
};

bool parse_placement(const char* spec, Placement* out) {
    // SIDE:PRICE:QTY:AFTER[:DISPLAY[:CANCEL_AFTER]]
    char side = 0;
    long long price = 0;
    long long qty = 0;
    long long after = 0;
    long long display = 0;
    long long cancel_after = 0;
    const int n = std::sscanf(spec, "%c:%lld:%lld:%lld:%lld:%lld", &side, &price, &qty,
                              &after, &display, &cancel_after);
    if (n < 4) return false;
    out->side = (side == 'B' || side == 'b') ? Side::Buy : Side::Sell;
    out->price = static_cast<Price4>(price);
    out->quantity = static_cast<uint32_t>(qty);
    out->after = static_cast<uint64_t>(after);
    out->display = (n >= 5) ? static_cast<uint32_t>(display) : 0;
    out->cancel_after = (n >= 6) ? static_cast<uint64_t>(cancel_after) : 0;
    return out->quantity > 0 && out->price > 0;
}

}  // namespace

int main(int argc, char** argv) {
    const char* feed = nullptr;
    const char* fills_path = nullptr;
    const char* json_path = nullptr;
    std::vector<Placement> places;

    for (int i = 1; i < argc; ++i) {
        const std::string a = argv[i];
        if (a == "--place" && i + 1 < argc) {
            Placement pl;
            if (!parse_placement(argv[++i], &pl)) {
                std::fprintf(stderr, "error: bad --place spec '%s'\n", argv[i]);
                return 2;
            }
            pl.id = 1000000 + places.size();
            places.push_back(pl);
        } else if (a == "--fills" && i + 1 < argc) {
            fills_path = argv[++i];
        } else if (a == "--json" && i + 1 < argc) {
            json_path = argv[++i];
        } else if (feed == nullptr) {
            feed = argv[i];
        } else {
            std::fprintf(stderr, "error: unexpected argument '%s'\n", argv[i]);
            return 2;
        }
    }
    if (feed == nullptr || places.empty()) {
        std::fprintf(stderr,
                     "usage: %s <feed.gz> "
                     "--place SIDE:PRICE:QTY:AFTER[:DISPLAY[:CANCEL_AFTER]] ...\n"
                     "                    [--fills out.csv] [--json out.json]\n",
                     argv[0]);
        return 2;
    }

    std::vector<Lane> lanes;
    for (Model m : {Model::Naive, Model::Optimistic, Model::Mbo, Model::Pessimistic}) {
        QueueConfig c;
        c.model = m;
        lanes.emplace_back(m, c);
    }

    Handler h;
    h.lanes = &lanes;
    h.places = &places;

    try {
        Reader reader(feed);
        parse(reader, h);
    } catch (const std::exception& e) {
        std::fprintf(stderr, "error: %s\n", e.what());
        return 1;
    }

    if (fills_path != nullptr) {
        std::FILE* f = std::fopen(fills_path, "w");
        if (f == nullptr) {
            std::fprintf(stderr, "error: cannot write %s\n", fills_path);
            return 1;
        }
        std::fprintf(f, "model,order_id,ts,side,price,shares,trigger\n");
        for (const Lane& l : lanes) {
            for (const SimFill& s : l.fills) {
                std::fprintf(f, "%s,%" PRIu64 ",%" PRIu64 ",%c,%" PRId32 ",%u,%s\n",
                             to_string(l.model), s.order_id, s.ts, to_char(s.side),
                             s.price, s.shares, trigger_name(s.trigger));
            }
        }
        std::fclose(f);
    }

    std::printf("%-12s %8s %10s %10s %8s\n", "model", "fills", "shares", "clamped", "anomaly");
    std::printf("------------------------------------------------------\n");
    for (const Lane& l : lanes) {
        uint64_t shares = 0;
        for (const SimFill& s : l.fills) shares += s.shares;
        std::printf("%-12s %8zu %10" PRIu64 " %10" PRIu64 " %8" PRIu64 "\n",
                    to_string(l.model), l.fills.size(), shares, l.q.clamp_events(),
                    l.q.priority_anomalies());
    }

    if (json_path != nullptr) {
        std::FILE* f = std::fopen(json_path, "w");
        if (f == nullptr) return 1;
        std::fprintf(f, "{\n  \"messages\": %" PRIu64 ",\n  \"models\": {", h.index);
        bool first = true;
        for (const Lane& l : lanes) {
            uint64_t shares = 0;
            for (const SimFill& s : l.fills) shares += s.shares;
            std::fprintf(f,
                         "%s\n    \"%s\": {\"fills\": %zu, \"shares\": %" PRIu64
                         ", \"clamp_events\": %" PRIu64 ", \"priority_anomalies\": %" PRIu64 "}",
                         first ? "" : ",", to_string(l.model), l.fills.size(), shares,
                         l.q.clamp_events(), l.q.priority_anomalies());
            first = false;
        }
        std::fprintf(f, "\n  }\n}\n");
        std::fclose(f);
    }
    return 0;
}

File: python/reference/queue_sim.py
"""Slow, obvious queue-position simulator — the Phase 6 oracle.

Mirrors include/itchbook/sim/queue_model.hpp decision for decision. The two must
produce byte-identical fill logs; where they disagree, one is wrong and it is
usually not this one. That discipline caught real bugs in phases 2 and 3.

The order of operations per message is load-bearing and matches the C++ driver:

  1. session/halt state first, so all lanes see identical eligibility
  2. pre-mutation snapshot (E/C/X/D/U carry only an order reference)
  3. mutate the book
  4. resolve and commit to every lane
  5. price priority against the post-mutation book
"""
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from reference import parser as itch          # noqa: E402
from reference.book import Book               # noqa: E402

NAIVE, OPTIMISTIC, MBO, PESSIMISTIC = "naive", "optimistic", "mbo", "pessimistic"
MODELS = (NAIVE, OPTIMISTIC, MBO, PESSIMISTIC)

TRADE, CANCEL, ADD, NONE = "trade", "cancel", "add", "none"


def at_or_through(side, limit, px):
    return px <= limit if side == "B" else px >= limit


def strictly_through(side, limit, px):
    return px < limit if side == "B" else px > limit


def shares_at(book, side, price):
    level = book.bids if side == "B" else book.asks
    return sum(book.orders[r]["shares"] for r in level.get(price, ()))


def refs_at(book, side, price):
    level = book.bids if side == "B" else book.asks
    return list(level.get(price, ()))


def resolve(m, pre):
    """A decoded message plus the pre-mutation state becomes a queue event."""
    r = {"type": m["type"], "cls": NONE, "known": False, "ref": 0,
         "side": "B", "resting_price": 0, "shares": 0,
         "printable": False, "print_price": 0, "ts": m["ts"]}
    t = m["type"]

    if t in ("A", "F"):
        r.update(cls=ADD, side=m["side"], resting_price=m["price"], shares=m["shares"])
    elif t == "E":
        if pre is None:
            return r
        r.update(cls=TRADE, known=True, ref=m["ref"], side=pre["side"],
                 resting_price=pre["price"], shares=m["shares"],
                 printable=True, print_price=pre["price"])
    elif t == "C":
        if pre is None:
            return r
        # THE rule: a C fills only when printable AND printing at the resting
        # price. A non-printable C is already reported elsewhere as part of an
        # aggregate cross print, and a repriced C matched under a regime this
        # model does not describe.
        cls = TRADE if (m["printable"] and m["price"] == pre["price"]) else CANCEL
        r.update(cls=cls, known=True, ref=m["ref"], side=pre["side"],
                 resting_price=pre["price"], shares=m["shares"],
                 printable=m["printable"], print_price=m["price"])
    elif t == "X":
        if pre is None:
            return r
        r.update(cls=CANCEL, known=True, ref=m["ref"], side=pre["side"],
                 resting_price=pre["price"], shares=m["shares"])
    elif t in ("D", "U"):
        # A delete removes whatever was left, which only the pre-mutation book
        # knows. A U is delete-then-add; the add half joins the back and has no
        # effect on anyone already resting.
        if pre is None:
            return r
        ref = m["ref"] if t == "D" else m["orig_ref"]
        r.update(cls=CANCEL, known=True, ref=ref, side=pre["side"],
                 resting_price=pre["price"], shares=pre["shares"])
    elif t == "P":
        # Hidden liquidity never sat in the displayed FIFO, so known stays False
        # and it cannot advance a queue. Its side field is deliberately unused.
        r.update(cls=TRADE, known=False, shares=m["shares"], printable=True,
                 print_price=m["price"], resting_price=m["price"])
    elif t == "Q":
        r.update(cls=NONE, shares=m["shares"], printable=True,
                 print_price=m["price"], resting_price=m["price"])
    return r


class Entry:
    def __init__(self, oid, side, price, quantity, display, arrived_ns):
        self.id = oid
        self.side = side
        self.price = price
        self.display = display if 0 < display < quantity else quantity
        self.slice_size = self.display
        self.hidden = quantity - self.display
        self.ahead = 0
        self.ahead0 = 0
        self.arrived_ns = arrived_ns
        self.refreshes = 0
        self.live = True


class QueueModel:
    def __init__(self, model, clamp=True, naive_fills_hidden=True):
        self.model = model
        self.clamp = clamp
        self.naive_fills_hidden = naive_fills_hidden
        self.entries = []
        self.ahead_sets = {}
        self.fills = []
        self.clamp_events = 0
        self.priority_anomalies = 0
        self.tradable = False

    # ---- our orders ----

    def place(self, book, oid, side, price, quantity, display, arrival_ns):
        e = Entry(oid, side, price, quantity, display, arrival_ns)
        e.ahead = shares_at(book, side, price) + self._own_ahead(side, price, arrival_ns)
        e.ahead0 = e.ahead
        self.entries.append(e)
        if self.model == MBO:
            self.ahead_sets[oid] = {
                r: book.orders[r]["shares"] for r in refs_at(book, side, price)
            }

    def _own_ahead(self, side, price, arrived_ns):
        return sum(o.display for o in self.entries
                   if o.live and o.display > 0 and o.side == side
                   and o.price == price and o.arrived_ns < arrived_ns)

    # ---- arithmetic ----

    def _trade_class(self, e, removed, ts, trigger, book):
        used = min(removed, e.ahead)
        e.ahead -= used
        overflow = removed - used
        if overflow == 0:
            return
        fill = min(overflow, e.display)     # capped at the DISPLAYED slice
        if fill == 0:
            return
        e.display -= fill
        e.ahead = 0
        self.fills.append({"model": self.model, "order_id": e.id, "ts": ts,
                           "side": e.side, "price": e.price, "shares": fill,
                           "trigger": trigger})
        if e.display == 0 and e.hidden > 0:
            self._refresh(e, book)

    @staticmethod
    def _cancel_class(e, removed, from_ahead):
        if from_ahead:
            e.ahead -= min(removed, e.ahead)

    def _refresh(self, e, book):
        slice_size = min(e.hidden, e.slice_size)
        e.hidden -= slice_size
        e.display = slice_size
        e.refreshes += 1
        e.ahead = (shares_at(book, e.side, e.price)
                   + self._own_ahead(e.side, e.price, e.arrived_ns))

    # ---- commit ----

    def commit(self, r, book):
        if not self.tradable:
            return
        for e in self.entries:
            if not e.live or e.display == 0:
                continue
            if self.model == NAIVE:
                self._naive(e, r, book)
            else:
                self._queued(e, r, book)
        if self.clamp:
            self._clamp(book)
        self._reap()

    def _opposite_reaches_us(self, e, book):
        best = book.best_ask() if e.side == "B" else book.best_bid()
        return best is not None and at_or_through(e.side, e.price, best)

    def _queued(self, e, r, book):
        # R3: a printable trade strictly through our limit. Side is checked only
        # when the message carries a reliable one, and it stands down while R1 is
        # live so the two mechanisms cannot claim the same flow.
        if (r["cls"] == TRADE and r["printable"]
                and strictly_through(e.side, e.price, r["print_price"])
                and (not r["known"] or r["side"] == e.side)
                and not self._opposite_reaches_us(e, book)):
            self._trade_class(e, r["shares"], r["ts"], "through", book)
            return
        if not r["known"] or r["side"] != e.side or r["resting_price"] != e.price:
            return
        if r["cls"] == TRADE:
            self._trade_class(e, r["shares"], r["ts"], "execution", book)
            if self.model == MBO:
                self._note_trade(e, r)
        elif r["cls"] == CANCEL:
            self._cancel_class(e, r["shares"], self._cancel_is_ahead(e, r))

    def _cancel_is_ahead(self, e, r):
        if self.model == OPTIMISTIC:
            return True
        if self.model == PESSIMISTIC:
            return False
        return r["ref"] in self.ahead_sets.get(e.id, {})

    def _note_trade(self, e, r):
        s = self.ahead_sets.get(e.id)
        if s is None:
            return
        if r["ref"] not in s:
            if e.ahead > 0:
                self.priority_anomalies += 1
            return
        if r["shares"] >= s[r["ref"]]:
            del s[r["ref"]]
        else:
            s[r["ref"]] -= r["shares"]

    def _naive(self, e, r, book):
        if r["cls"] != TRADE or not r["printable"]:
            return
        if r["type"] == "Q":
            return
        if r["type"] == "P" and not self.naive_fills_hidden:
            return
        if not at_or_through(e.side, e.price, r["print_price"]):
            return
        fill = min(r["shares"], e.display)
        if fill == 0:
            return
        e.display -= fill
        self.fills.append({"model": self.model, "order_id": e.id, "ts": r["ts"],
                           "side": e.side, "price": e.price, "shares": fill,
                           "trigger": "naive"})
        if e.display == 0 and e.hidden > 0:
            self._refresh(e, book)

    def apply_price_priority(self, book, ts):
        """R1: the other side came to our price. Identical in all four models."""
        if not self.tradable:
            return
        for e in self.entries:
            if not e.live or e.display == 0:
                continue
            best = book.best_ask() if e.side == "B" else book.best_bid()
            if best is None or not at_or_through(e.side, e.price, best):
                continue
            other = "S" if e.side == "B" else "B"
            reach = 0
            for price, _shares, _count in book.top(other, 32):
                if not at_or_through(e.side, e.price, price):
                    break
                reach += _shares
            if reach == 0:
                continue
            self._trade_class(e, reach, ts, "lock", book)
        self._reap()

    def _clamp(self, book):
        for e in self.entries:
            if not e.live or e.display == 0:
                continue
            limit = (shares_at(book, e.side, e.price)
                     + self._own_ahead(e.side, e.price, e.arrived_ns))
            if e.ahead > limit:
                self.clamp_events += 1
                e.ahead = limit

    def _reap(self):
        for e in self.entries:
            if e.live and e.display == 0 and e.hidden == 0:
                e.live = False


def run(path, placements):
    """placements: list of (side, price, quantity, after, display)."""
    lanes = {m: QueueModel(m) for m in MODELS}
    book = Book()
    index = 0
    placed = [False] * len(placements)

    opener = __import__("gzip").open if str(path).endswith(".gz") else open
    with opener(path, "rb") as f:
        for mtype, payload in itch.iter_messages(f):
            m = itch.decode(mtype, payload)
            if m is None:
                continue

            # 1. session and halt state
            if m["type"] == "H":
                for q in lanes.values():
                    q.tradable = m["state"] == "T"
            elif m["type"] == "S":
                if m["code"] == "Q":
                    for q in lanes.values():
                        q.tradable = True
                elif m["code"] in ("M", "C"):
                    for q in lanes.values():
                        q.tradable = False

            # 2. pre-mutation snapshot
            pre = None
            ref = m.get("orig_ref") if m["type"] == "U" else m.get("ref")
            if m["type"] in ("E", "C", "X", "D", "U") and ref is not None:
                o = book.orders.get(ref)
                if o is not None:
                    pre = {"side": o["side"], "price": o["price"], "shares": o["shares"]}

            # 3. mutate
            book.apply(m)

            # 4. resolve and commit
            r = resolve(m, pre)
            for q in lanes.values():
                q.commit(r, book)

            # 5. price priority
            for q in lanes.values():
                q.apply_price_priority(book, m["ts"])

            index += 1
            for i, (side, price, qty, after, display) in enumerate(placements):
                if placed[i] or index < after:
                    continue
                for q in lanes.values():
                    q.place(book, 1000000 + i, side, price, qty, display, m["ts"])
                placed[i] = True

    return lanes, index
